core/src/main/scala/spinal/core/FixedPoint.scala

```diff
@@ -135,8 +138,1437 @@ object getFixSym{
+          aligned = f.expand(0, this.fracWidth-f.fracWidth)
+        }
+        // Resize the integer portion
+        var resized: Fix = aligned
+        if (f.wholeWidth > this.wholeWidth) {
+          if (f.hasTag(tagSaturate)) {
+            resized = aligned.sat(f.wholeWidth-this.wholeWidth)
+          } else {
+            resized = this.clone
+            resized.setName(s"${this.getName()}_resized")
+            resized.raw := aligned.raw.takeHigh(this.signWidth) ## aligned.raw.takeLow(this.numericWidth)
+          }
+        } else if (f.wholeWidth < this.wholeWidth) {
+          resized = aligned.expand(this.wholeWidth-f.wholeWidth)
+        }
+        this.raw := resized.raw
+      case b: Bits => this := b
+      case s: SInt => this := s
+      case u: UInt => this := u
+      case _ => SpinalError("Undefined assignment")
+    }
+  }
+
+  override def toString: String = s"${component.getPath() + "/" + this.getDisplayName()} : ${getClass.getSimpleName}[signed=${if (this.signed) "1" else "0"},whole=${this.wholeWidth},frac=${this.fracWidth}]"
+}
+
+object AutoFix {
+
+  def apply(f: Fix): AutoFix = {
+    val maxValue = BigInt(2).pow(f.bitWidth-1)-1
+    val minvalue = -BigInt(2).pow(f.bitWidth-1)
+    val ret = new AutoFix(maxValue, minvalue, -f.fracWidth exp)
+    ret.raw := f.raw
+    ret
+  }
+
+  def apply(num: BigInt, exp: ExpNumber): AutoFix = {
+    val ret = new AutoFix(num, num, exp)
+    if (num >= 0)
+      ret.raw := BigIntToUInt(num).asBits
+    else
+      ret.raw := BigIntToSInt(num).asBits
+    ret
+  }
+
+}
+
+class AutoFix(val maxValue: BigInt, val minValue: BigInt, val exp: ExpNumber) extends MultiData {
+//  assert((maxValue*BigDecimal(2).pow(-exp.value)).isWhole,
+//    s"maxValue ${maxValue} is not representable with exponent 2^${exp.value}")
+//  assert((minValue*BigDecimal(2).pow(-exp.value)).isWhole,
+//    s"maxValue ${minValue} is not representable with exponent 2^${exp.value}")
+
+  // TODO?: Support dropping sign bit iff max and min have the same sign?
+  private val needsSign = (maxValue < 0) ^ (minValue < 0)
+
+  val signed = (maxValue < 0) || (minValue < 0)
+  val signWidth = if (signed) 1 else 0
+
+  private val maxShifted = maxValue.abs - (if (maxValue < 0) signWidth else 0)
+  private val maxBits = maxShifted.bitLength
+  private val minShifted = minValue.abs - (if (minValue < 0) signWidth else 0)
+  private val minBits = minShifted.bitLength
+
+  val bitWidth = Math.max(maxBits, minBits) + signWidth
+  val fracWidth = Math.min(exp.value, 0)
+  val wholeWidth = bitWidth - fracWidth - signWidth
+  val intWidth = bitWidth - fracWidth
+
+  val raw: Bits = Bits(bitWidth bit)
+
+  raw.setRefOwner(this)
+  raw.setPartialName("", weak = true)
+
+  override def elements: ArrayBuffer[(String, Data)] = {
+    ArrayBuffer("" -> raw)
+  }
+
+  // These are used for reference tracking
+  private var parents: Seq[AutoFix] = Nil
+  private val children: mutable.Set[AutoFix] = mutable.Set()
```

Comment:
For which purpose is this children ?

Reply:
The thought was that in the future it might be used to graph out a chain of operations for analysis. I can remove it. It has no purpose at the moment.

Reply:
Ahhh, no worries, we can keep it ^^